File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/LocalSearchHandler.java
package com.company.proto.handlers;

import com.company.proto.Utils;
import com.company.proto.torrent.Torrent;
import com.google.protobuf.ByteString;

import java.util.Map;
import java.util.regex.Pattern;
import java.util.regex.PatternSyntaxException;

public class LocalSearchHandler implements Handler {
	private Map<String, ByteString> storage;
	
	public LocalSearchHandler(Map<String, ByteString> storage) {
		this.storage = storage;
	}
	
	private static boolean isValid(String regex) {
		try {
			Pattern.compile(regex);
			return true;
		} catch (PatternSyntaxException exception) {
			return false;
		}
	}
	
	@Override
	public Torrent.Message handle(Torrent.Message message) {
		String regex = message.getLocalSearchRequest().getRegex();
		
		if (!isValid(regex)) return messageError(regex);
		
		String resultFilename = storage.keySet().stream()
				.filter(filename -> filename.matches(regex))
				.findAny()
				.orElse("");
		if (resultFilename.isEmpty()) return successNoResult(regex);
		else return successWithResult(resultFilename);
	}
	
	private Torrent.Message successWithResult(String resultFilename) {
		byte[] fileData = storage.get(resultFilename).toByteArray();
		Torrent.FileInfo fileInfo = Torrent.FileInfo.newBuilder()
				.setFilename(resultFilename)
				.setHash(Utils.hashToMD5(fileData))
				.setSize(fileData.length)
				.addAllChunks(Utils.toList(fileData))
				.build();
		Torrent.LocalSearchResponse searchResponse = Torrent.LocalSearchResponse
				.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.setErrorMessage("SUCCESS")
				.addFileInfo(fileInfo)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.setLocalSearchResponse(searchResponse)
				.build();
	}
	
	private Torrent.Message successNoResult(String filename) {
		Torrent.FileInfo fileInfo = Torrent.FileInfo.newBuilder()
				.setFilename(filename)
				.build();
		Torrent.LocalSearchResponse searchResponse = Torrent.LocalSearchResponse
				.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.setErrorMessage("SUCCESS")
				.addFileInfo(fileInfo)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.setLocalSearchResponse(searchResponse)
				.build();
	}
	
	private Torrent.Message messageError(String filename) {
		Torrent.FileInfo fileInfo = Torrent.FileInfo.newBuilder()
				.setFilename(filename).build();
		Torrent.LocalSearchResponse searchResponse = Torrent.LocalSearchResponse
				.newBuilder()
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("MESSAGE_ERROR")
				.addFileInfo(fileInfo)
				.build();
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.setLocalSearchResponse(searchResponse)
				.build();
	}
}

File: GPU-Distributed/Torrent/src/main/java/com/company/proto/handlers/LocalSearchHandler.kt
package com.company.proto.handlers

import com.company.proto.*
import com.company.proto.torrent.Torrent
import com.google.protobuf.ByteString

class LocalSearchHandler(private val storage: Map<Torrent.FileInfo, ByteString>) : Handler {
	
	override fun handle(message: Torrent.Message): Torrent.Message {
		val regex = message.localSearchRequest.regex
		
		if (!regex.isValidRegex()) return messageError(regex)
		
		val filesInfo = storage.keys.filter { regex.toRegex().matches(it.filename) }
		println("Files found $filesInfo")
		if (filesInfo.isEmpty()) return successNoResult()
		
		return successWithResult(filesInfo)
	}
	
	private fun successWithResult(fileInfo: List<Torrent.FileInfo>): Torrent.Message {
		val localSearchResponse = Torrent.LocalSearchResponse.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.addAllFileInfo(fileInfo)
				.build()
		return Torrent.Message.newBuilder()
				.setLocalSearchResponse(localSearchResponse)
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.build()
	}
	
	private fun successNoResult(): Torrent.Message {
		val searchResponse = Torrent.LocalSearchResponse.newBuilder()
				.setStatus(Torrent.Status.SUCCESS)
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.setLocalSearchResponse(searchResponse)
				.build()
	}
	
	private fun messageError(regex: String): Torrent.Message {
		val searchResponse = Torrent.LocalSearchResponse.newBuilder()
				.setStatus(Torrent.Status.MESSAGE_ERROR)
				.setErrorMessage("Regex $regex is invalid")
				.build()
		return Torrent.Message.newBuilder()
				.setType(Torrent.Message.Type.LOCAL_SEARCH_RESPONSE)
				.setLocalSearchResponse(searchResponse)
				.build()
	}
}

